# 06: Regression

If a linear relationship exists between two variables, then one variable can be predicted from the other by using the least squares regression equation. 

- **Least Squares Regression Equation**: The equation that minimizes the total of all squared prediction errors for known Y scores in the original correlation analysis. 

$$
Y^\prime = bX + a
$$

- $Y^\prime$ is the predicted value. 
- $X$ represents the known value. 
- $b$ is the slope of the line 
- $a$ is the Y-intercept 

$$
b = r\sqrt{\dfrac{SS_y}{SS_x}}
$$

- $SS_y$ is the sum of squares for all Y scores 
- $SS_x$ is the sum of squares for all x scores 

$$
a = \bar{Y} - b\bar{X}
$$

In [1]:
import math

In [2]:
def ss(X): 
    m = sum(X)/len(X)
    return sum([(x - m)**2 for x in X])

In [3]:
def spxy(X, Y): 
     mx = sum(X)/len(X)
     my = sum(Y)/len(Y)
     XY = []
     for i in range(len(X)):
          XY.append((X[i], Y[i]))
     return sum((x-mx)*(y-my) for x,y in XY)

In [4]:
def pearson_r(X, Y): 
    return spxy(X, Y) / (math.sqrt(ss(X)*ss(Y)))

In [5]:
def calc_b(r, ssy, ssx): 
    return r*math.sqrt(ssy/ssx)

In [6]:
def calc_a(mx, my, b): 
    return my - mx*b

In [7]:
def regression(x, b, a): 
    return (b*x) + a

#### Example 01

Assume that an r of .30 describes the relationship between educational level (highest grade completed) and estimated number of hours spent reading each week. More specifically:

X = Education Level 

Y = Weekly Reading Time 

$$
\bar{X} = 13, \quad
\bar{Y} = 8, \quad
SS_x = 25, \quad
SS_y = 50, \quad
r = 0.30
$$



1. Determine the least squares equation for predicting weekly reading time from educational level.

In [8]:
b = calc_b(r=0.3, ssy=50, ssx=25)
a = calc_a(mx=13, my=8, b=b)

round(a, 4), round(b, 4)

(2.4846, 0.4243)

$$
Y^\prime = 0.4243X + 2.4846
$$

2. Faith’s education level is 15. What is her predicted reading time?

In [10]:
Y = regression(x=15, a=a, b=b)

print(f"Faith's predicted reading time is {Y:.3f} hours.")

Faith's predicted reading time is 8.849 hours.


3. Keegan’s educational level is 11. What is his predicted reading time?

In [11]:
Y = regression(x=11, a=a, b=b)

print(f"Keegan's predicted reading time is {Y:.3f} hours.")

Keegan's predicted reading time is 7.151 hours.


## Standard Error 

- **Standard Error of Estimate ($S_{y|x}$)**: A rough measure of the average amount of predictive error. 

$$
S_{y|x} = \sqrt{\dfrac{SS_{y|x}}{n - 2}} = \sqrt{\dfrac{\sum(Y - Y^\prime)^2}{n-2}} = \sqrt{\dfrac{SS_y(1-r)^2}{n-2}}
$$ 

- It is a special kind of standard deviation that reflects the magnitude of predictive error. 

- Because any straight line, including the regression line, can be made to coincide with two data points, 2 degrees of freedom are lost, hence the denominator $n-2$.

In [12]:
def std_err(Y=[], r=0, ssy = False, n = False): 
    ssy = ssy if ssy else ss(Y)
    n = n if n else len(Y)
    return math.sqrt((ssy*(1-(r**2)))/(n-2))

### Assumptions 

1. Linearity - use of regression equation requires that the underlying relationship be linear. 

2. Homoscedasticity - use of the standard error of estimate, $s_{x|y}$, assumes that except for chance, the dots in the original scatterplot will be dispersed equally about all segments of the regression line. 


## Interpretation of $r^2$

- **Squared Correlation Coefficient ($r^2$)**: The proportion of the total variability in one variable that is predictable from its relationship with  the other variable. 

- Primitive effort is to predict the $\bar{Y}$ repeatedly for each value of $Y$. This provides a frame of reference against which to evaluate predictive effort based on correlation. 

- Errors should be smaller when customized predictions of $Y^\prime$ from the least squares equations can be used compared to when only repetitive prediction of $\bar{Y}$ is used. 

Gain in accuracy = $SS_y - SS_{y|x}$

$$
r^2 = \dfrac{SS_{y^\prime}}{SS_y} = \dfrac{SS_y - SS_{y|x}}{SS_y}
$$

- $r^2$ is the proportion by which the variability of all $Y$ scores, as measured by $SS_y$ can be reduced. 

    - $r^2$ of 0.01 reflects weak relationship 
    - $r^2$ of 0.09 reflects moderate relationship 
    - $r^2$ of 0.25 reflects strong relationship 

- $r^2$ doesn't ensure cause-effect 

### Multiple Regression Equations 

- **Multiple Regression Equation**: A least squares equation that contains more than one predictor or X variable. 

**Example**
$$
Y^\prime = .41(X_1) + .005(X_2) + 0.001(X_3) + 1.03
$$

Where $X_1, X_2, X_3$ are predictors of $Y^\prime$(criterion variable)



### Regression Toward the Mean
- **Regression Toward the Mean**: A tendency for scores, particularly extreme scores, to shrink toward the mean. It occurs for individuals or subsets, but not entire groups. 

- **Regression Fallacy**: Occurs whenever regression toward the mean is interpreted as a real, rather than a change, effect. 